File: samples/feed-demo-ios/SellwildSample/LegacyScreen.swift
import SwiftUI
import UIKit
import SellwildSDK

/// Legacy: the deprecated WebView widget. It runs Prebid.js in a WebView and
/// cannot earn the CPMs native ads do, so it is only on this screen.
struct LegacyScreen: View {
    let config: SellwildConfig
    @State private var status = "Loading the widget"

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            VStack(alignment: .leading, spacing: 4) {
                Text("Legacy WebView widget (deprecated)")
                    .font(.title3.bold())
                    .accessibilityIdentifier(SampleID.legacyTitle)
                Text("SellwildWidgetView. Use the Feed and Ads screens instead.")
                    .font(.footnote)
                    .foregroundColor(.secondary)
                Text(status)
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .accessibilityIdentifier(SampleID.legacyStatus)
            }
            .padding(16)
            LegacyWidget(config: config) { status = $0 }
        }
    }
}

/// `SellwildWidgetView` (UIKit) in SwiftUI, with its delegate.
struct LegacyWidget: UIViewRepresentable {
    let config: SellwildConfig
    let onStatus: (String) -> Void

    func makeCoordinator() -> Coordinator { Coordinator(onStatus: onStatus) }

    func makeUIView(context: Context) -> SellwildWidgetView {
        let widget = SellwildWidgetView(config: config)
        widget.delegate = context.coordinator
        widget.accessibilityIdentifier = SampleID.legacyWebView
        widget.load()
        return widget
    }

    func updateUIView(_ widget: SellwildWidgetView, context: Context) {}

    final class Coordinator: SellwildWidgetViewDelegate {
        private let onStatus: (String) -> Void

        init(onStatus: @escaping (String) -> Void) {
            self.onStatus = onStatus
        }

        func sellwildWidgetViewDidLoad(_ widgetView: SellwildWidgetView) {
            onStatus("Widget loaded")
        }

        func sellwildWidgetView(_ widgetView: SellwildWidgetView, didTapListing listing: SellwildListing) {
            guard let link = listing.url, let url = URL(string: link) else { return }
            UIApplication.shared.open(url)
        }

        func sellwildWidgetView(_ widgetView: SellwildWidgetView, didFailWithError error: Error) {
            onStatus("Widget error: \(error.localizedDescription)")
        }
    }
}

File: samples/feed-demo-android/app/src/main/java/com/sellwild/sample/LegacyScreen.kt
package com.sellwild.sample

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import com.sellwild.sdk.SellwildConfig
import com.sellwild.sdk.SellwildListing
import com.sellwild.sdk.SellwildWidgetView

/**
 * Legacy: the deprecated WebView widget. It runs Prebid.js in a WebView and cannot earn
 * the CPMs native ads do, so it is only on this screen. Do not copy it into a new app.
 */
@Composable
fun LegacyScreen(config: SellwildConfig) {
    var status by remember { mutableStateOf("Loading the widget") }
    val holder = remember { ViewHolder<SellwildWidgetView>() }
    PauseResumeEffect(onPause = { holder.view?.pause() }, onResume = { holder.view?.resume() })
    Column(Modifier.fillMaxSize()) {
        Column(Modifier.fillMaxWidth().padding(16.dp)) {
            Text(
                "Legacy WebView widget (deprecated)",
                Modifier.testTag(SampleIds.LEGACY_TITLE),
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
            )
            Text(
                "SellwildWidgetView. Use the Feed and Ads screens instead.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Text(
                status,
                Modifier.testTag(SampleIds.LEGACY_STATUS),
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        // The id is on a Box around the widget: a test tag on an AndroidView does not
        // reach UI Automator.
        Box(Modifier.fillMaxSize().testTag(SampleIds.LEGACY_WEBVIEW)) {
            LegacyWidget(config, holder) { status = it }
        }
    }
}

/** `SellwildWidgetView` with its listener. [onStatus] gets its load and error lines. */
@Composable
private fun LegacyWidget(config: SellwildConfig, holder: ViewHolder<SellwildWidgetView>, onStatus: (String) -> Unit) {
    val status by rememberUpdatedState(onStatus)
    AndroidView(
        factory = { context ->
            SellwildWidgetView(context).apply {
                listener = object : SellwildWidgetView.Listener {
                    override fun onWidgetLoaded(widgetView: SellwildWidgetView) {
                        status("Widget loaded")
                    }

                    override fun onListingTapped(listing: SellwildListing) {
                        val url = listing.url ?: return
                        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
                    }

                    override fun onError(widgetView: SellwildWidgetView, message: String) {
                        status("Widget error: $message")
                    }
                }
                setup(config)
                load()
                holder.view = this
            }
        },
        modifier = Modifier.fillMaxSize(),
        onRelease = { it.destroy() },
    )
}
